lib/Dialect/SPIRV/Transforms/DecorateWithLayoutInfoSPIRVPass.cpp

```diff
@@ -0,0 +1,282 @@
+    return type;
+  }
+  switch (type.getKind()) {
+  case spirv::TypeKind::Struct:
+    return SPIRVLayoutInfoUtils::decorateType(type.cast<spirv::StructType>(),
+                                              offset);
+  case spirv::TypeKind::Array:
+  case StandardTypes::Vector:
+    return SPIRVLayoutInfoUtils::decorateType(type.cast<spirv::CompositeType>(),
+                                              offset);
+  case spirv::TypeKind::Pointer:
+    return SPIRVLayoutInfoUtils::decorateType(type.cast<spirv::PointerType>(),
+                                              offset);
+  default:
+    llvm_unreachable("unhandled SPIR-V type");
+  }
+}
+
+Type SPIRVLayoutInfoUtils::decorateType(spirv::PointerType pointerType,
+                                  spirv::StructType::LayoutInfo &offset) {
+  uint64_t currentOffset = 0;
+  auto pointeeType = SPIRVLayoutInfoUtils::decorateType(
+      pointerType.getPointeeType(), currentOffset);
+  // According to the SPIR-V spec the size of the pointer depends on the
+  // addressing model. So, it should be save to increment an offset for the
+  // pointer by 8 bytes, because 8 bytes is the upper bound limit for the
+  // pointer size.
+  offset += 8;
+  return spirv::PointerType::get(pointeeType, pointerType.getStorageClass());
+}
+
+Type SPIRVLayoutInfoUtils::decorateType(spirv::CompositeType compositeType,
+                                  spirv::StructType::LayoutInfo &offset) {
+  if (!compositeType.getNumElements()) {
+    return compositeType;
+  }
+  uint64_t elementSizeInBytes = 0;
+  auto memberType = SPIRVLayoutInfoUtils::decorateType(
+      compositeType.getElementType(0), elementSizeInBytes);
+  offset += elementSizeInBytes * compositeType.getNumElements();
+  return SPIRVLayoutInfoUtils::getArrayOrVectorType(compositeType, memberType,
+                                                    elementSizeInBytes);
+}
+
+Type SPIRVLayoutInfoUtils::getArrayOrVectorType(
+    spirv::CompositeType compositeType, Type elementType,
+    uint64_t arrayStride) {
+  switch (compositeType.getKind()) {
+  case spirv::TypeKind::Array:
+    return spirv::ArrayType::get(elementType, compositeType.getNumElements(),
+                                 arrayStride);
+  case StandardTypes::Vector:
+    return VectorType::get(compositeType.getNumElements(), elementType);
+  default:
+    llvm_unreachable("unhandled SPIR-V composite type");
+  }
+}
+
+unsigned SPIRVLayoutInfoUtils::getPrimitiveTypeSizeInBytes(Type primitiveType) {
+  // According to the SPIR-V spec:
+  // "The ArrayStride, MatrixStride, and Offset decorations must be large
+  // enough to hold the size of the objects they affect (that is, specifying
+  // overlap is invalid)."
+  auto bitWidth = primitiveType.getIntOrFloatBitWidth();
+  if (bitWidth == 1 || bitWidth == 8 || bitWidth == 16)
+    return 4;
+  return bitWidth / 8;
+}
+
+bool SPIRVLayoutInfoUtils::isLegalType(Type type) {
```

Comment:
Could you add some comments of what is legal and what is not. Even if just a copy of paste from relevant portion of the spec, thats fine.